lib/ElementHandle.js

```diff
@@ -137,24 +140,67 @@ class ElementHandle extends JSHandle {
     return {x, y, width, height};
   }
 
+  /**
+   * @return {!Promise<?{x: number, y: number, width: number, height: number}>}
+   */
+  async _assertBoundingBox() {
+    const boundingBox = await this.boundingBox();
+    if (boundingBox)
+      return boundingBox;
+
+    const noElementError = await this.executionContext().evaluate(element => {
+      if (!element.isConnected)
+        return 'Node is detached from document';
+      if (element.nodeType !== Node.ELEMENT_NODE)
+        return 'Node is not of type HTMLElement';
+      return 'Node is not visible';
+    }, this);
+
+    throw new Error(noElementError);
+  }
+
   /**
    *
    * @param {!Object=} options
    * @returns {!Promise<Object>}
    */
   async screenshot(options = {}) {
-    await this._scrollIntoViewIfNeeded();
+    let needsViewportReset = false;
+
+    let boundingBox = await this._assertBoundingBox();
+
+    const viewport = this._page.viewport();
+
+    if (boundingBox.width > viewport.width || boundingBox.height > viewport.height) {
```

Comment:
can we move this functionality into `page.screenshot` so that it works with clips larger then viewport? Seems reasonable to me.

Reply:
that might be a bit tricky, because resizing the viewport will likely move the bounding box to an unknown location.

Reply:
I think solving that in `page.screenshot` should be a separate PR, but resizing the viewport in `page.screenshot` seems like it would do bad things more often than `elementHandle.screenshot` does, since at the page level, many things are based on the viewport size. If possible, that should be solved in chromium/protocol.

I'd be good with scrolling in `page.screenshot` if the clip area requires it, but I'd like to get this PR merged first.